File: param-bounds-vectors-run.py
#!/usr/bin/env python3
"""
CLC-v1 §6.5 extended parameter-bound vectors runner (rev CLC-1.10) - reads
param-bounds-vectors.json and runs them against the Python entails().

Asserts BOTH the verdict and the resolved reason code (canonical, code before
the first ':').  kind=param-defaults applies the scheme-default rule first.
"""
import json
import os
import sys

from clc_semantics import entails, materialize_defaults


def canonical_reason(s):
    if not s:
        return ""
    return s.split(":", 1)[0] if ":" in s else s


def main():
    path = os.environ.get("CLC_PARAM_BOUNDS_VECTORS") or os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "..", "capability", "data", "_vectors", "clc-v1", "param-bounds-vectors.json")
    with open(path) as f:
        vectors = json.load(f)

    pass_count = 0
    fail_count = 0
    for v in vectors:
        grant = v.get("grant") or {}
        op = v.get("request") or {}
        if v["kind"] == "param-defaults":
            op = materialize_defaults(grant, op, v.get("scheme_defaults"))
        r = entails(grant, op)
        got = "allow" if r["entails"] else "deny"
        got_reason = canonical_reason(r.get("reason", ""))
        want = v["expect"]["verdict"]
        want_reason = v["expect"].get("reason") or ""
        ok = got == want and got_reason == want_reason
        if ok:
            pass_count += 1
        else:
            fail_count += 1
            print("%-10s FAIL want=%s/%s got=%s/%s derivation=%s" % (
                v["id"], want, want_reason, got, got_reason, v["derivation"]))

    print("Total: %d | Pass: %d | Fail: %d" % (len(vectors), pass_count, fail_count))
    return 1 if fail_count else 0


if __name__ == "__main__":
    sys.exit(main())

File: ts/param-bounds-vectors-run.ts
// CLC-v1 §6.5 extended parameter-bound vectors runner (rev CLC-1.10).
// Reads param-bounds-vectors.json and runs them against entails(); asserts the
// verdict and the resolved reason.  kind=param-defaults applies defaults first.
import { readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { entails, materializeDefaults } from './clc_semantics.ts';
import type { Grant, Operation } from './clc_semantics.ts';

interface Expectation { verdict: string; reason?: string }
interface Vector {
    id: string;
    kind: string;
    grant?: Grant;
    request?: Operation;
    scheme_defaults?: Record<string, unknown>;
    expect: Expectation;
    derivation: string;
}
function canonicalReason(s?: string): string {
    if (!s) return '';
    const i = s.indexOf(':');
    return i >= 0 ? s.slice(0, i) : s;
}
const here = dirname(fileURLToPath(import.meta.url));
const path = process.env.CLC_PARAM_BOUNDS_VECTORS
    || resolve(join(here, '..', '..', 'capability', 'data', '_vectors', 'clc-v1', 'param-bounds-vectors.json'));
const vectors: Vector[] = JSON.parse(readFileSync(path, 'utf8'));
let pass = 0;
let fail = 0;
for (const v of vectors) {
    const grant = v.grant ?? { id: '' };
    let op = v.request ?? { id: '' };
    if (v.kind === 'param-defaults') {
        op = materializeDefaults(grant, op, v.scheme_defaults);
    }
    const r = entails(grant, op);
    const got = r.entails ? 'allow' : 'deny';
    const gotReason = canonicalReason(r.reason);
    const wantReason = v.expect.reason ?? '';
    if (got === v.expect.verdict && gotReason === wantReason) {
        pass++;
    } else {
        fail++;
        console.log(`${v.id} FAIL want=${v.expect.verdict}/${wantReason} got=${got}/${gotReason} ${v.derivation}`);
    }
}
console.log(`Total: ${vectors.length} | Pass: ${pass} | Fail: ${fail}`);
if (fail) process.exit(1);
